### PR TITLE
x-pack/metricbeat/module/gcp: Fix missing ECS labels when using exclude_labels: true

### LINKED ISSUE #40437: GCP metrics: Missing ECS labels when using `exclude_labels: true`
The GCP `metrics` metricset's `exclude_labels: true` option causes an issue in the `groupTimeSeries` function. The default metadata collector is created as a value, preventing the function from initializing `metadataService` with the correct time series data. This results in missing important labels like `cloud.instance.id` in the events. To fix this, `metadataService` should be initially set to `nil`, allowing initialization based on the time series parameter.

```go
var metadataService gcp.MetadataService
```

### PR DESCRIPTION
**Pull request by Linu-Elias**
## Proposed commit message
The GCP `metrics` metricset's `exclude_labels: true` option causes an issue in the `groupTimeSeries` function. The default metadata collector is created as a value, preventing the function from initializing `metadataService` with the correct time series data. This results in missing important labels like` cloud.instance.id` in the events. To fix this, `metadataService` should be initially set to nil, allowing initialization based on the time series parameter.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes #40437 

## Screenshots

**Compute Service Before Fix:**
```json
{
  "_index": ".ds-metricbeat-8.16.0-2024.08.12-000001",
  "_id": "zWdtRpEBVgxWUufvh8Eh",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-08-12T11:46:00.000Z",
    "metricset": {
      "name": "compute",
      "period": 60000
    },
    "gcp": {
      "labels": {},
      "compute": {
        "instance": {
          "uptime_total": {
            "sec": 965207
          },
          "memory": {
            "balloon": {
              "ram_size": {
                "value": 4115619840
              },
              "swap_in": {
                "bytes": 0
              },
              "swap_out": {
                "bytes": 0
              },
              "ram_used": {
                "value": 1657262080
              }
            }
          },
          "uptime": {
            "sec": 60.000000000116415
          },
          "cpu": {
            "usage": {
              "pct": 0.11856303568638395
            },
            "usage_time": {
              "sec": 7.113782141183037
            }
          },
          "network": {
            "egress": {
              "packets": {
                "count": 1217
              },
              "bytes": 883471
            },
            "ingress": {
              "packets": {
                "count": 980
              },
              "bytes": 230375
            }
          },
          "disk": {
            "read": {
              "bytes": 0
            },
            "write_ops_count": {
              "value": 1230
            },
            "read_ops_count": {
              "value": 0
            },
            "write": {
              "bytes": 12386800
            }
          }
        }
      }
    },
    "service": {
      "type": "gcp"
    },
    "agent": {
      "name": "EPINPUNW06A8",
      "type": "metricbeat",
      "version": "8.16.0",
      "ephemeral_id": "6882ed99-deee-4fae-a6ee-af617960c5ea",
      "id": "f49fbc8d-cc6a-4629-89b1-454f33f48d82"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "host": {},
    "event": {
      "duration": 684338677,
      "dataset": "gcp.compute",
      "module": "gcp"
    }
  }
}
```

**Compute Service After fix:** 
```json
{
  "_index": ".ds-metricbeat-8.16.0-2024.08.12-000001",
  "_id": "uGc7RpEBVgxWUufvzb4e",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-08-12T10:52:00.000Z",
    "metricset": {
      "name": "compute",
      "period": 60000
    },
    "gcp": {
      "labels": {},
      "compute": {
        "instance": {
          "uptime_total": {
            "sec": 1980
          },
          "cpu": {
            "usage_time": {
              "sec": 0.13933221539459595
            },
            "usage": {
              "pct": 0.011611017949549662
            }
          },
          "uptime": {
            "sec": 60
          }
        }
      }
    },
    "service": {
      "type": "gcp"
    },
    "cloud": {
      "region": "us-central1",
      "instance": {
        "id": "8449630262774775859"
      },
      "account": {
        "id": "elastic-obs-integrations-dev",
        "name": "elastic-obs-integrations-dev"
      },
      "provider": "gcp",
      "availability_zone": "us-central1-a"
    },
    "host": {},
    "ecs": {
      "version": "8.0.0"
    },
    "agent": {
      "ephemeral_id": "c1f3c9a5-ff55-4d65-a586-e751805d1347",
      "id": "f49fbc8d-cc6a-4629-89b1-454f33f48d82",
      "name": "EPINPUNW06A8",
      "type": "metricbeat",
      "version": "8.16.0"
    },
    "event": {
      "duration": 706928630,
      "dataset": "gcp.compute",
      "module": "gcp"
    }
  }
}
```

**Spanner Service Before Fix:**

```json
{
  "_index": ".ds-metricbeat-8.16.0-2024.08.12-000001",
  "_id": "CTKeRpEBGeZDOp8bEh6T",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-08-12T12:40:46.200Z",
    "event": {
      "duration": 921850497,
      "dataset": "gcp.metrics",
      "module": "gcp"
    },
    "metricset": {
      "name": "metrics",
      "period": 60000
    },
    "gcp": {
      "metrics": {
        "instance": {
          "leader_percentage_by_region": {
            "value": 1
          }
        }
      },
      "labels": {}
    },
    "service": {
      "type": "gcp"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "host": {},
    "agent": {
      "type": "metricbeat",
      "version": "8.16.0",
      "ephemeral_id": "4227a7c2-941d-45b0-ab3d-0328c7a3ee97",
      "id": "f49fbc8d-cc6a-4629-89b1-454f33f48d82",
      "name": "EPINPUNW06A8"
    }
  }
}
```

**Spanner Service After Fix:**

```json
{
  "_index": ".ds-metricbeat-8.16.0-2024.08.12-000001",
  "_id": "fDK1RpEBGeZDOp8bDx-J",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-08-12T13:06:00.000Z",
    "host": {},
    "agent": {
      "ephemeral_id": "00deca2b-9a95-42af-809c-730759f69cdf",
      "id": "f49fbc8d-cc6a-4629-89b1-454f33f48d82",
      "name": "EPINPUNW06A8",
      "type": "metricbeat",
      "version": "8.16.0"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "cloud": {
      "provider": "gcp",
      "instance": {
        "id": "spanner-1"
      },
      "account": {
        "id": "elastic-obs-integrations-dev",
        "name": "elastic-obs-integrations-dev"
      }
    },
    "event": {
      "dataset": "gcp.metrics",
      "module": "gcp",
      "duration": 955200812
    },
    "metricset": {
      "name": "metrics",
      "period": 60000
    },
    "gcp": {
      "metrics": {
        "instance": {
          "cpu": {
            "utilization_by_operation_type": {
              "value": 0
            }
          }
        }
      },
      "labels": {
        "metrics": {
          "priority": "medium",
          "database": "db-1",
          "category": "misc",
          "is_system": "false"
        },
        "resource": {
          "instance_config": "regional-europe-west1",
          "location": "europe-west1"
        }
      }
    },
    "service": {
      "type": "gcp"
    }
  }
}
```

### CODE REVIEW

**Comment by gpop63** [comment]
@Linu-Elias please add 1 event before & after this change in JSON format so we can see the impact of this change.

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -170,6 +170,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - Fix missing metrics from CloudWatch when include_linked_accounts set to false. {issue}40071[40071] {pull}40135[40135]
 - Update beat module with apm-server monitoring metrics fields {pull}40127[40127]
 - Fix Azure Monitor metric timespan to restore Storage Account PT1H metrics {issue}40376[40376] {pull}40367[40367]
+- Fix GCP missing ECS labels when using exclude_labels: true {issue}40437[40437] {pull}40467[40467]
```
Comment (gpop63) [inline CHANGELOG.next.asciidoc:173]:
```suggestion
- Fix missing ECS Cloud fields in GCP `metrics` metricset when using `exclude_labels: true` {issue}40437[40437] {pull}40467[40467]
```

Comment (Linu-Elias) [reply CHANGELOG.next.asciidoc]: Done

**Review by kaiyan-sheng — APPROVED** [review]
I would prefer some tests added for this. But @gpop63 mentioned it's not a small task. Could you open a followup issue to add some tests please? Thank you!

**Review by gpop63 — APPROVED** [review]

**Review by shmsr — APPROVED** [review]

**Pull request merged**